nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/registry/reflect/MarshallerInspector.java

```diff
@@ -303,20 +310,24 @@ private RenderingContext getRealContext(RenderingContext ctx) {
      * @since 7.2
      */
     private Object getSingletonInstance(RenderingContext ctx) {
-        if (singleton == null) {
-            singleton = getNewInstance(ctx, true);
-        } else {
-            for (Field contextField : contextFields) {
-                ThreadSafeRenderingContext value;
-                try {
-                    value = (ThreadSafeRenderingContext) contextField.get(singleton);
-                } catch (IllegalArgumentException | IllegalAccessException e) {
-                    log.error("unable to create a marshaller instance for clazz " + clazz.getName(), e);
-                    return null;
+        if (!servicesInjected) {
+            synchronized (this) {
+                if (!servicesInjected) {
+                    injectServices(singleton);
+                    servicesInjected = true;
```

Comment:
The `injectServices` method already does `servicesInjected = true` so this is redundant.
But there are other callers than this one and they aren't called with a synchronized(this) that I can see to protect the `servicesInjected` update. This should be made consistent.